### ISSUE TITLE
BUG: read_sql with query doesn't get metadata from the database

### ISSUE THREAD

**Issue by mvashishtha** [pandas-dev/pandas#46221]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd

filename = "/tmp/sqlite_for_modin.db"
con = "sqlite:///{}".format(filename)
# Create a 1 x 1 dataframe of floats where the only entry is null
input = pd.DataFrame([[None]]).astype("float")
# The only dtype is float64
print(f"input dtypes: {input.dtypes}")

# Write the dataframe to the database in the temporary file
input.to_sql("test_1x1_null", con, index=False, if_exists="replace")

# Read the whole table with a query
output = pd.read_sql("SELECT * FROM test_1x1_null", con)
# The only dtype is object
print(f"output dtypes: {output.dtypes}")
```

### Issue Description

I want to use `read_sql` to get the metadata of a SQL table. In this example, I write a dataframe of floats, with all elements null, to a table. When I run the `sqlite3` CLI and get the schema of the table with

```
.open /tmp/sqlite_for_modin.db;
pragma table_info('test_1x1_null');
```

I get 

```
0|0|FLOAT|0||0
```

but when I read the first row from the table with pandas, my metadata says nothing about floats. Instead of the original float64 types:

```
0   float64
dtype: object
```
I get:

```
0   object
dtype: object
```

I believe that pandas is trying to infer dtypes from the data it reads rather than using the query result's schema, which hopefully it should be able to access using the query and the connection. Note that `read_sql_table` can infer types even when the whole table is null, e.g. if use `read_sql_table` instead of `read_sql` with a query in the example:

```python
import pandas as pd

filename = "/tmp/sqlite_for_modin.db"
con = "sqlite:///{}".format(filename)
# Create a 1 x 1 dataframe of floats where the only entry is null
input = pd.DataFrame([[None]]).astype("float")
# The only dtype is float64
print(f"input dtypes: {input.dtypes}")

# Write the dataframe to the database in the temporary file
input.to_sql("test_1x1_null", con, index=False, if_exists="replace")

# Read the whole table with read_sql_table
output = pd.read_sql("test_1x1_null", con)
# The only dtype is object
print(f"output dtypes: {output.dtypes}")
```

I get the correct dtypes.

### Expected Behavior

In the example above, the result of `read_sql` should have dtypes `float64`. I want pandas to infer dtypes using the schema from the database rather than trying to infer them on its own. 

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 06d230151e6f18fdb8139d09abf539867a8cd481
python           : 3.9.10.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 21.3.0
Version          : Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.1
numpy            : 1.22.1
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 22.0.3
setuptools       : 60.8.2
Cython           : 0.29.27
pytest           : 7.0.0
hypothesis       : None
sphinx           : 4.4.0
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.8.0
html5lib         : None
pymysql          : None
psycopg2         : 2.9.3
jinja2           : 3.0.3
IPython          : 8.0.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
fsspec           : 2022.02.0
gcsfs            : None
matplotlib       : 3.5.1
numba            : None
numexpr          : 2.8.1
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : 0.16.0
pyarrow          : 6.0.1
pyreadstat       : None
pyxlsb           : None
s3fs             : 2022.02.0+2.g61f99be
scipy            : 1.7.3
sqlalchemy       : 1.4.31
tables           : 3.7.0
tabulate         : None
xarray           : 0.20.2
xlrd             : 2.0.1
xlwt             : None
zstandard        : None

</details>

**Comment by ParfaitG** [comment]
As you show, issue may come down to reading a table vs reading a query. The former carries metadata on types while the latter does not since it is a generated resultset at runtime. 

Since SQL queries can contain calculated expressions and even typecasting that differs from original source(s), all columns of a query are really evaluated expressions derived from input but no longer adheres to input specs. 

Therefore, when an engine queries a column of all NULLs, it would return a type-less column where the DB-API returns all `None` types and in turn pandas reads as `object`.

**Comment by mvashishtha** [comment]
@ParfaitG I agree that a query can have metadata that's very different from any table in the database. However, the sqlalchemy solution [here](https://stackoverflow.com/a/65373459/17554722) should allow pandas to infer the types of the query result for any PEP-249-compliant database. Some extra code would be needed to convert from type codes to [type objects](https://www.python.org/dev/peps/pep-0249/#type-objects), and from those to pandas types.

**Comment by devin-petersohn** [comment]
I wonder if it's possible to avoid the expensive type inference that pandas does at the end if we can just use the database schema as provided by sqlalchemy. In my opinion, the pandas dtypes should match (as close as possible) the schema of the database. It should also be faster if we can determine the types a priori. 

I'm willing to help out here. I think there's a good case to be made for better performance, correctness, and determinism.

**Comment by fraghag** [comment]
Seems like there should be another dtype_backend option for infering dtypes from the db metadata.
